shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java

```diff
@@ -230,39 +251,31 @@ private void addPointerForIndex(
 
     packet.putLong(motionEventId); // motionEventId
     packet.putLong(timeStamp); // time_stamp
-    packet.putLong(pointerChange); // change
-    packet.putLong(pointerKind); // kind
+    if (ongoingPans.containsKey(event.getPointerId(pointerIndex))) {
```

Comment:
Extract this recurring condition as a boolean.

Reply:
Or do you think it makes sense to extract the trackpad related data to another function? Because it seems that there are  contains exclusive logic on both side, such as adding to and removing from `ongoingPans`, and many `if` clauses.

If we define a "data packet" class (as commented by a TODO on L207 of this file), then we don't have to worry about assignment order or packet structure. We can assign the common fields first, then split exclusive parts into two functions.

Reply:
Just extracted the condition for now. It would be nice to use a data packet if it could be used in the test class as well. I don't know that much about how the java code interoperates with the rest of the engine, but I assume we can't change the signature on FlutterRenderer::dispatchPointerDataPacket ?

Reply:
The "data packet" class is actually just a builder. Check out https://github.com/flutter/engine/blob/main/lib/ui/window/pointer_data_packet.h .